modernization overhaul

ExpandingMan opened the pull request:
I have made the following changes:
- DataStreams and DataFrames are required.  This is because I was having problems with `@require` when using JDBC in other packages, so basically the DataFrames functionality would have been largely useless.  This seems very important and useful functionality for this package, so I'm hoping that you'll agree with this choice.  The overall dependency list is still relatively small.
- I have removed the dependence on the abandoned DBAPI.jl package and updated the DBAPI code in JDBC to be standalone and to provide a minimal "Julian" interface (or at least a minimal layer on top of the direct Java calls).  I have renamed some things to make a little more sense in this new context.
- Tests have been updated and modernized.
- Should be pretty much all ready for 0.7.  I haven't tested on 0.7 yet, as JavaCall will undoubtedly require extensive updates.  As part of this, all references to `Nullable` have been removed, and `Union{T, Nothing}` is used instead.  This should not cause any performance issues on 0.6 because it is only used for some constructors and some initialization checks and never in pulling data.

I know I have made a lot of changes here, some of which will be breaking (but quite minor), but the package seemed to be floundering so I thought it was in need of an overhaul.  I have found JDBC to be the most reliable and easy-to-setup solution for querying a wide variety of databases in Julia, and I've used it successfully without any significant memory issues on datasets as big as 10 GB without significant memory issues.

I'd also be happy to update the README, but let's wait to see if you are ok with these changes before I do that.

aviks commented:
Can't we just do a dependency on DataStreams? Do we need to depend on DataFrames as well? Other than that, I'm happy with all the changes.

ExpandingMan commented:
Ok, I've done the following: DataFrames is no longer a dependency.  Instead I've added `JDBC.load` methods, which can be called like e.g. `JDBC.load(DataFrame, csr)`.  In principle this can work with other types depending on how they have the DataStreams interface implemented.  Anyway, this still allows us to retain the user-friendliness that we would have had with DataFrames as I expect it will be quite common to call `JDBC.load`.  I'm not sure how much I like the name `load` so I'm certainly open to suggestions if you want to change it.

Since you seem basically on-board I'm not overhauling the README.  Please wait until I've updated docs to merge! (will let you know...)

ExpandingMan commented:
Ok!  I have added some tests and cleaned up the README.  We still need better documentation, but I think this gets the job done.  Please let me know of any additional changes you'd like to see.  Thanks!

Considering what seems to be the low usage of this package, I'd like to ask that this be tagged as a release relatively soon, and I'd like to make an announcement on discourse about it.  I have noticed on discourse that quite a few people seem to be struggling with finding convenient database interfaces in Julia, and indeed I was suffering myself until I realized how relatively easy this package is to use.  I suspect once people become a bit more aware this package will start to see significantly more use.

aviks commented:
Thanks for taking this on. I agree this is the best way to get data out of a database into Julia at this time, but I might be biased :)

That kind of a datastream interface is exactly what I had in mind, thanks!

I've invited to you to be a collaborator, so if you accept, you can make a release yourself. Let me know if still want me to do it. I agree, a discourse post would be good, particularly since there might be minor incompatibilities from earlier versions.

ExpandingMan commented:
Great, thanks!  I just tagged the release.

Yeah, the situation is awkward right now, because ODBC.jl probably performs much better because it uses a proper C API, and the performance on JDBC.jl is probably quite bad.  It has proven *very* easy to set up though, and I've even done some pulls of several GB so the performance can't be *that* bad...
